lib/semantics/resolve-labels.cc

```diff
@@ -0,0 +1,969 @@
+  }
+  bool SwScope() {
+    PopScope();
+    return PushScope();
+  }
+
+  template<typename A> bool PushName(const A &a) {
+    const auto &optionalName{std::get<0>(std::get<0>(a.t).statement.t)};
+    if (optionalName.has_value()) {
+      constructNames_.push_back(optionalName.value().ToString());
+    }
+    return PushScope();
+  }
+  bool PushName(const parser::BlockConstruct &blockConstruct) {
+    const auto &optionalName{
+        std::get<parser::Statement<parser::BlockStmt>>(blockConstruct.t)
+            .statement.v};
+    if (optionalName.has_value()) {
+      constructNames_.push_back(optionalName.value().ToString());
+    }
+    return PushScope();
+  }
+  template<typename A> bool PushNonBlockName(const A &a) {
+    const auto &optionalName{std::get<0>(std::get<0>(a.t).statement.t)};
+    if (optionalName.has_value()) {
+      constructNames_.push_back(optionalName.value().ToString());
+    }
+    return true;
+  }
+
+  template<typename A> void PopNonBlockConstructName(const A &a) {
+    CheckName(a);
+    SelectivePopBack(a);
+  }
+
+  template<typename A> void SelectivePopBack(const A &a) {
+    const auto &optionalName{std::get<0>(std::get<0>(a.t).statement.t)};
+    if (optionalName.has_value()) {
+      constructNames_.pop_back();
+    }
+  }
+  void SelectivePopBack(const parser::BlockConstruct &blockConstruct) {
+    const auto &optionalName{
+        std::get<parser::Statement<parser::BlockStmt>>(blockConstruct.t)
+            .statement.v};
+    if (optionalName.has_value()) {
+      constructNames_.pop_back();
+    }
+  }
+
+  /// \brief Check constraints and pop scope
+  template<typename A> void PopName(const A &a) {
+    CheckName(a);
+    PopScope();
+    SelectivePopBack(a);
+  }
+
+  /// \brief Check <i>case-construct-name</i> and pop the scope
+  /// Constraint C1144 - opening and ending name must match if present, and
+  /// <i>case-stmt</i> must either match or be unnamed
+  void PopName(const parser::CaseConstruct &caseConstruct) {
+    CheckName(caseConstruct, "CASE");
+    PopScope();
+    SelectivePopBack(caseConstruct);
+  }
+
+  /// \brief Check <i>select-rank-construct-name</i> and pop the scope
+  /// Constraints C1154, C1156 - opening and ending name must match if present,
+  /// and <i>select-rank-case-stmt</i> must either match or be unnamed
+  void PopName(const parser::SelectRankConstruct &selectRankConstruct) {
+    CheckName(selectRankConstruct, "RANK", "RANK ");
+    PopScope();
+    SelectivePopBack(selectRankConstruct);
+  }
+
+  /// \brief Check <i>select-construct-name</i> and pop the scope
+  /// Constraint C1165 - opening and ending name must match if present, and
+  /// <i>type-guard-stmt</i> must either match or be unnamed
+  void PopName(const parser::SelectTypeConstruct &selectTypeConstruct) {
+    CheckName(selectTypeConstruct, "TYPE", "TYPE ");
+    PopScope();
+    SelectivePopBack(selectTypeConstruct);
+  }
+
+  // -----------------------------------------------
```

Comment:
We generally don't expend screen space on horizontal rules.

Reply:
deleted